src/Root.cpp

```diff
@@ -127,156 +124,190 @@ void cRoot::Start(std::unique_ptr<cSettingsRepositoryInterface> overridesRepo)
-			#endif
+		#endif
 
-			while (!m_ShouldStop && !m_bRestart && !m_TerminateEventRaised)  // These are modified by external threads
+		for (;;)
+		{
+			m_StopEvent.Wait();
+
+			if (m_TerminateEventRaised && m_RunAsService)
 			{
-				std::this_thread::sleep_for(std::chrono::seconds(1));
+				// Dont kill if running as a service
+				m_TerminateEventRaised = false;
 			}
-
-			if (m_TerminateEventRaised)
+			else
 			{
-				m_ShouldStop = true;
+				break;
 			}
+		}
 
-			// Stop the server:
-			m_WebAdmin->Stop();
+		// Stop the server:
+		m_WebAdmin->Stop();
 
-			LOG("Shutting down server...");
-			m_Server->Shutdown();
-		}  // if (m_Server->Start())
-		else
-		{
-			m_ShouldStop = true;
-		}
+		LOG("Shutting down server...");
+		m_Server->Shutdown();
+	}  // if (m_Server->Start()
 
-		delete m_MojangAPI; m_MojangAPI = nullptr;
+	delete m_MojangAPI; m_MojangAPI = nullptr;
 
-		LOGD("Shutting down deadlock detector...");
-		dd.Stop();
+	LOGD("Shutting down deadlock detector...");
+	dd.Stop();
 
-		LOGD("Stopping world threads...");
-		StopWorlds();
+	LOGD("Stopping world threads...");
+	StopWorlds();
 
-		LOGD("Stopping authenticator...");
-		m_Authenticator.Stop();
+	LOGD("Stopping authenticator...");
+	m_Authenticator.Stop();
 
-		LOGD("Freeing MonsterConfig...");
-		delete m_MonsterConfig; m_MonsterConfig = nullptr;
-		delete m_WebAdmin; m_WebAdmin = nullptr;
+	LOGD("Freeing MonsterConfig...");
+	delete m_MonsterConfig; m_MonsterConfig = nullptr;
+	delete m_WebAdmin; m_WebAdmin = nullptr;
 
-		LOGD("Unloading recipes...");
-		delete m_FurnaceRecipe;   m_FurnaceRecipe = nullptr;
-		delete m_CraftingRecipes; m_CraftingRecipes = nullptr;
+	LOGD("Unloading recipes...");
+	delete m_FurnaceRecipe;   m_FurnaceRecipe = nullptr;
+	delete m_CraftingRecipes; m_CraftingRecipes = nullptr;
 
-		LOG("Unloading worlds...");
-		UnloadWorlds();
+	LOG("Unloading worlds...");
+	UnloadWorlds();
 
-		LOGD("Stopping plugin manager...");
-		delete m_PluginManager; m_PluginManager = nullptr;
+	LOGD("Stopping plugin manager...");
+	delete m_PluginManager; m_PluginManager = nullptr;
 
-		cItemHandler::Deinit();
+	cItemHandler::Deinit();
 
-		LOG("Cleaning up...");
-		delete m_Server; m_Server = nullptr;
+	LOG("Cleaning up...");
+	delete m_Server; m_Server = nullptr;
 
-		LOG("Shutdown successful!");
-	}
+	m_InputThreadRunFlag.clear();
+	#ifdef _WIN32
+		DWORD Length;
+		INPUT_RECORD Record
+		{
+			KEY_EVENT,
+			{
+				{
+					TRUE,
+					1,
+					VK_RETURN,
+					MapVirtualKey(VK_RETURN, MAPVK_VK_TO_VSC),
+					{ { VK_RETURN } },
+					0
+				}
+			}
+		};
 
+		// Can't kill the input thread since it breaks cin (getline doesn't block / receive input on restart)
+		// Apparently no way to unblock getline
+		// Only thing I can think of for now
+		if (WriteConsoleInput(GetStdHandle(STD_INPUT_HANDLE), &Record, 1, &Length) == 0)
+		{
+			LOGWARN("Couldn't notify the input thread; the server will hang before shutdown!");
```

Comment:
I think this is of sufficient importance to notify the user, since it isn't supposed to happen and could cause confusion since the server won't exit as expected.